webviz_subsurface/plugins/_well_completions.py

```diff
@@ -0,0 +1,473 @@
+        kh_values.append(kh_real)
+    return compl_events, kh_values
+
+
+def format_time_series(
+    open_frac: list, shut_frac: list, kh_mean: list, kh_min: list, kh_max: list
+) -> Optional[Dict]:
+    """
+    The functions takes in five lists with values per timestep
+    * fractions of realizations open in this zone
+    * fractions of realizations shut in this zone
+    * kh mean over open realizations
+    * kh min over open realizations
+    * kh max over open realizations
+
+    Returns the data in compact form:
+    {
+        t: [3, 5],
+        open: [0.25, 1.0],
+        shut: [0.75, 0.0],
+        khMean: [500, 1000],
+        khMin: [200, 800],
+        khMean: [600, 1500]
+    }
+    """
+    output: Dict[str, list] = {
+        "t": [],
+        "open": [],
+        "shut": [],
+        "khMean": [],
+        "khMin": [],
+        "khMax": [],
+    }
+    prev_open_val = prev_shut_val = 0
+
+    for (
+        i,
+        (open_val, shut_val, kh_mean_val, kh_min_val, kh_max_val),
+    ) in enumerate(zip(open_frac, shut_frac, kh_mean, kh_min, kh_max)):
+        if open_val != prev_open_val or shut_val != prev_shut_val:
+            output["t"].append(i)
+            output["open"].append(open_val)
+            output["shut"].append(shut_val)
+            output["khMean"].append(kh_mean_val)
+            output["khMin"].append(kh_min_val)
+            output["khMax"].append(kh_max_val)
+        prev_open_val = open_val
+        prev_shut_val = shut_val
+
+    if not output["t"]:
+        return None
+    return output
+
+
+def calc_over_realizations(
+    compl_events: list, kh_values: list, realizations: list
+) -> tuple:
+    # pylint: disable=assignment-from-no-return
+    """
+    Takes in two three dimensional lists where the levels are: 1. realization \
+    2. zones and 3. timesteps
+
+    Returns two dimensional lists where calculations have been done over the \
+    realization level.
+    """
+    # calculate fraction of open realizations
+    open_count = np.maximum(np.asarray(compl_events), 0)  # remove -1
+    open_count_reduced = open_count.sum(axis=0)  # sum over realizations
+    open_frac = (
+        np.asarray(open_count_reduced, dtype=np.float64) / float(len(realizations))
+    ).round(decimals=3)
+
+    # calculate fraction of shut realizations
+    shut_count = (
+        np.minimum(np.asarray(compl_events), 0) * -1
+    )  # remove +1 and convert -1 to 1
+    shut_count_reduced = shut_count.sum(axis=0)  # sum over realizations
+    shut_frac = (
+        np.asarray(shut_count_reduced, dtype=np.float64) / float(len(realizations))
+    ).round(decimals=3)
+
+    # calculate khMean, khMin and khMax
+    kh_mean = (np.asarray(kh_values).sum(axis=0) / float(len(realizations))).round(
+        decimals=2
+    )
+    kh_min = np.asarray(kh_values).min(axis=0).round(decimals=2)
+    kh_max = np.asarray(kh_values).max(axis=0).round(decimals=2)
+
+    return open_frac, shut_frac, kh_mean, kh_min, kh_max
```

Comment:
There are some duplications of `np.asarray(variable)` here. Worth creating local variables to not duplicate the `np.asarray` calls?

Reply:
yes, good idea